Content.Shared/Storage/EntitySystems/SharedStorageSystem.cs

```diff
@@ -589,28 +582,89 @@ public bool PlayerInsertHeldEntity(EntityUid uid, EntityUid player, StorageCompo
     /// <param name="player">The player to insert an entity with</param>
+    /// <param name="toInsert"></param>
     /// <returns>true if inserted, false otherwise</returns>
-    public bool PlayerInsertEntityInWorld(EntityUid uid, EntityUid player, EntityUid toInsert, StorageComponent? storageComp = null)
+    public bool PlayerInsertEntityInWorld(Entity<StorageComponent?> uid, EntityUid player, EntityUid toInsert)
     {
-        if (!Resolve(uid, ref storageComp) || !_sharedInteractionSystem.InRangeUnobstructed(player, uid))
+        if (!Resolve(uid, ref uid.Comp) || !_sharedInteractionSystem.InRangeUnobstructed(player, uid))
             return false;
 
-        if (!Insert(uid, toInsert, out _, user: player, storageComp))
+        if (!Insert(uid, toInsert, out _, user: player, uid.Comp))
         {
             _popupSystem.PopupClient(Loc.GetString("comp-storage-cant-insert"), uid, player);
             return false;
         }
         return true;
     }
 
+    /// <summary>
+    /// Returns true if there is enough space to theoretically fit another item.
+    /// </summary>
+    public bool HasSpace(Entity<StorageComponent?> uid)
+    {
+        if (!Resolve(uid, ref uid.Comp))
+            return false;
+
+        return GetCumulativeItemSizes(uid, uid.Comp) < GetMaxTotalWeight(uid) &&
+               uid.Comp.Container.ContainedEntities.Count < uid.Comp.MaxSlots;
+    }
+
+    /// <summary>
+    /// Returns the sum of all the ItemSizes of the items inside of a storage.
+    /// </summary>
+    public int GetCumulativeItemSizes(EntityUid uid, StorageComponent? component = null)
+    {
+        if (!Resolve(uid, ref component))
+            return 0;
+
+        var sum = 0;
+        foreach (var item in component.Container.ContainedEntities)
+        {
+            if (!_itemQuery.TryGetComponent(item, out var itemComp))
+                continue;
+            sum += SharedItemSystem.GetItemSizeWeight(itemComp.Size);
+        }
+
+        return sum;
+    }
+
+    public ItemSize GetMaxItemSize(Entity<StorageComponent?> uid)
+    {
+        if (!Resolve(uid, ref uid.Comp))
+            return DefaultStorageMaxItemSize;
+
+        // If we specify a max item size, use that
+        if (uid.Comp.MaxItemSize != null)
+            return uid.Comp.MaxItemSize.Value;
+
+        if (!TryComp<ItemComponent>(uid, out var item))
+            return DefaultStorageMaxItemSize;
+
+        // if there is no max item size specified, the value used
+        // is one below the item size of the storage entity, clamped at ItemSize.Tiny
+        return (ItemSize) Math.Max((int) item.Size - 1, 1);
+    }
+
+    public int GetMaxTotalWeight(Entity<StorageComponent?> uid)
+    {
+        if (!Resolve(uid, ref uid.Comp))
+            return 0;
+
+        if (uid.Comp.MaxTotalWeight != null)
+            return uid.Comp.MaxTotalWeight.Value;
+
+        return uid.Comp.MaxSlots * SharedItemSystem.GetItemSizeWeight(GetMaxItemSize(uid));
+    }
```

Comment:
AFAICT, if the max weight is not explicitly defined, then it will never be a limiting factor - the max item count will always block insertions.

Unless there'd ever be a situation where we'd want the automatic max total weight to be more restrictive, then all of the code that calls `GetMaxTotalWeight()` should just get wrapped in a `if (comp.MaxTotalWeight != null)` instead, as that'd avoid having to constantly compute the current total weight, which itself requires many component lookups.

Reply:
Or is the plan to eventually allow things to modify weight (i.e., the weight of a bag includes all contents). In which case this'd no longer be the case and you'd still have to always compute the total weight.

Reply:
At least in ss13 count can be a limiting factor if you put a lot of small stuff in though for now I think the goal is to get the system first and balance more after as changing all of the yaml is a large piece of work.

Reply:
it's an optional balancing factor that storages can opt to use.